### PR TITLE
RDRP-903: construction unit tests - construction_helpers.py

### PR DESCRIPTION
**Pull request by JenCheshire**
Grouped read/validate and post-construction tasks into separate functions.

Tweaks so NI will run

Unit tests for functions within construction_helpers.py

## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/construction/construction_read_validate.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+def read_validate_construction_files(
+    config: dict,
+    check_file_exists: Callable,
+    read_csv: Callable,
+    is_northern_ireland: bool = False,
+    run_construction: bool = False,
+    run_postcode_construction: bool = False,
+):
+    """Read and validate construction files the construction module."""
```
Comment (cbrowning04) [inline src/construction/construction_read_validate.py]: A full docstring here would be benefitial

File: src/construction/construction_read_validate.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+"""Read and validate construction files the construction module."""
+import logging
+import pandas as pd
+from typing import Callable
```
Comment (cbrowning04) [inline src/construction/construction_read_validate.py]: Reorder these imports

```python
import logging
from typing import Callable

import pandas as pd
```

File: src/construction/construction_helpers.py [hunk 3]
```diff
@@ -148,4 +149,45 @@ def clean_construction_type(value: str) -> str:
...
+
+def finalise_forms_gb(updated_snapshot_df: pd.DataFrame) -> pd.DataFrame:
+    """Tasks to prepare the GB forms for the next stage in pipeline.
+
+    Args:
+        updated_snapshot_df (pd.DataFrame): The updated snapshot df.
+
+    Returns:
+        pd.DataFrame: The updated snapshot df with postcodes_harmonised
+        and short forms reset.
```
Comment (cbrowning04) [inline src/construction/construction_helpers.py]: tab in

File: tests/test_construction/test_construction_helpers.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
+"""Tests for construction_helpers.py"""
+import pandas as pd
+import numpy as np
+from pandas._testing import assert_frame_equal
+import logging
+import pytest
+
```
Comment (cbrowning04) [inline tests/test_construction/test_construction_helpers.py]: import re-order (PEP8)
'.' at end of docstring for consistency

File: tests/test_construction/test_construction_helpers.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+def test__convert_formtype():
+    assert _convert_formtype("1") == "0001"
+    assert _convert_formtype("1.0") == "0001"
+    assert _convert_formtype("0001") == "0001"
+    assert _convert_formtype("6") == "0006"
+    assert _convert_formtype("6.0") == "0006"
+    assert _convert_formtype("0006") == "0006"
+    assert _convert_formtype("2") is None
+    assert _convert_formtype("") is None
+    assert _convert_formtype(None) is None
```
Comment (cbrowning04) [inline tests/test_construction/test_construction_helpers.py]: 1. Docstring for this test
2. Maybe a common assert message for all assert. e.g., 
`msg = {message}`
and then add this after each assertion

Comment (cbrowning04) [reply tests/test_construction/test_construction_helpers.py]: Also, what happens if the formtype passed is an int or float or of another type.

File: tests/test_construction/test_construction_helpers.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+            ["B", 202412, None, "Form sent out", "0006"],
+            ["C", 202412, 0, "Other status", "0001"],
+            ["D", 202412, 1, "Form sent out", "0006"],
+            ["E", 202412, None, "Other status", "0001"],
+        ]
+        input_snapshot_df = pd.DataFrame(data=data, columns=input_cols)
+        return input_snapshot_df
+
+    # Create construction df
+    def create_test_construction_df(self):
```
Comment (cbrowning04) [inline tests/test_construction/test_construction_helpers.py]: bit of a chore and not needed, but return types could be added to each of the 'fixtures' in all of the tests. Up to you.

File: tests/test_construction/test_construction_helpers.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+            snapshot_output.reset_index(drop=True), expected_snapshot_output
+        ), "Snapshot output is not as expected"
+
+
+def test_clean_construction_type():
+    assert clean_construction_type("new") == "new"
+    assert clean_construction_type("Short to Long") == "short_to_long"
+    assert clean_construction_type("  ") is np.NaN
+    assert clean_construction_type("") is np.NaN
+    assert clean_construction_type(None) is np.NaN
```
Comment (cbrowning04) [inline tests/test_construction/test_construction_helpers.py]: Same comments as similar styled tests above.

File: src/construction/construction_read_validate.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        # NI data has no instance but needs an instance of 1
+        if is_northern_ireland:
+            construction_df["instance"] = 1
+            construction_df["construction_type"] = None
+        if isinstance(construction_df, type(None)):
+            construction_df = pd.DataFrame()
+            return construction_df
+
+        else:
+            # validate and merge schemas
```
Comment (cbrowning04) [inline src/construction/construction_read_validate.py]: are you able to make this comment more accurate? I did it in another PR however this code has now been moved :)

**Review by cbrowning04 — CHANGES_REQUESTED** [review]
Great stuff Jen. I've made a few tiny styling+docstring changes and pushed them, so feel free to look over them.
Regarding comments on this PR, they're mostly small changes, and the main functionalities are implemented well, so only some small comments.
Thanks @JenCheshire

File: tests/test_construction/test_construction_helpers.py (hunk 5 above)
Comment (JenCheshire) [reply tests/test_construction/test_construction_helpers.py]: formtype is converted to string in the pipeline, but I've added a line to ensure, and added to test

**Review by cbrowning04 — APPROVED** [review]

**Pull request merged**